python/eight_mile/pytorch/layers.py

```diff
@@ -1807,15 +1894,98 @@ def forward(self, qkvm: Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Te
         return self.w_O(x)
 
 
+class MultiHeadedRelativeAttention(nn.Module):
+    """
+    Multi-headed relative attention from Shaw et al 2018 (https://www.aclweb.org/anthology/N18-2074.pdf)
+
+    This method follows the same approach of MultiHeadedAttention, but it computes Relative Position Representations (RPR)
+    which are used as part of the attention computations.  To facilitate this, the model has its own internal
+    embeddings lookup table, and it has an updated computation for both the attention weights and the application
+    of those weights to follow them.
+
+    """
+    def __init__(self, num_heads: int, d_model: int, rpr_k: int, dropout: float = 0.1, scale: bool = False, d_k: Optional[int] = None):
+        """Constructor for multi-headed attention
+
+        :param h: The number of heads
+        :param d_model: The model hidden size
+        :param dropout (``float``): The amount of dropout to use
+        :param scale: Should we scale the dot product attention
+        :param d_k: The low-order project per head.  This is normally `d_model // num_heads` unless set explicitly
+        """
+        super().__init__()
+
+        if d_k is None:
+            self.d_k = d_model // num_heads
+            if d_model % num_heads != 0:
+                raise Exception(f"d_model ({d_model}) must be evenly divisible by num_heads ({num_heads})")
+        else:
+            self.d_k = d_k
+
+        self.rpr_k = rpr_k
+        self.rpr_key = nn.Embedding(2 * rpr_k + 1, self.d_k)
+        self.rpr_value = nn.Embedding(2 * rpr_k + 1, self.d_k)
+
+        self.h = num_heads
+        self.w_Q = Dense(d_model, self.d_k * self.h)
+        self.w_K = Dense(d_model, self.d_k * self.h)
+        self.w_V = Dense(d_model, self.d_k * self.h)
+        self.w_O = Dense(self.d_k * self.h, d_model)
+        if scale:
+            self.attn_fn = SeqScaledDotProductRelativeAttention(dropout)
+        else:
+            self.attn_fn = SeqDotProductRelativeAttention(dropout)
+        self.attn = None
+
+    def make_rpr(self, seq_len, device) -> Tuple[torch.Tensor, torch.Tensor]:
```

Comment:
Because the pytorch embeddings don't allow for negative numbers I see that you are shifting all the distances by `self.rpr_k` so that a distance of `-self.rpr_k` becomes `0` so the emebddings can handle it. I did the same thing in my SNLI implementation [here](https://github.com/blester125/baseline/blob/dbfdac7515dd67d37651701a4538a9679048686f/python/addons/NLI.py#L293) so that makes sense but there should be some comments to explain what you are doing. At first glance it looks like we are ignoring the negative positions